# Get the important Features

In this Notebook is analysed by a Frequency table **the most Extremely or desicive Positive and Negative Words**

Here is getting:

**x4, x5, x6 and x7**

In [1]:
import pandas as pd
# if you want to download the original file:
#df = pd.read_csv('https://raw.githubusercontent.com/rasbt/pattern_classification/master/data/50k_imdb_movie_reviews.csv')
# otherwise load local file
df = pd.read_csv('shuffled_movie_data.csv')
df.tail()

review  sentiment
49995  OK, lets start with the best. the building. al...          0
49996  The British 'heritage film' industry is out of...          0
49997  I don't even know where to begin on this one. ...          0
49998  Richard Tyler is a little boy who is scared of...          0
49999  I waited long to watch this movie. Also becaus...          1

In [2]:
# Return a lower case proccesed text
def processtextT(texto):
    import re
    REPLACE_NO_SPACE = re.compile("(\.)|(\;)|(\:)|(\!)|(\')|(\?)|(\,)|(\")|(\()|(\))|(\[)|(\])|(\*)(\n)")
    REPLACE_WITH_SPACE = re.compile("(<br\s*/><br\s*/>)|(\-)|(\/)")
    texto = REPLACE_NO_SPACE.sub('', texto.lower())
    texto = REPLACE_WITH_SPACE.sub(' ', texto)
    return texto

In [3]:
df2 = df.copy()

In [4]:
dataframe = pd.DataFrame(columns=['review', 'sentiment'])

In [5]:
dataframe

Empty DataFrame
Columns: [review, sentiment]
Index: []

In [7]:
for i, row in df2.iterrows():
    dataframe = dataframe.append({'review':processtextT(row['review']), 'sentiment':row['sentiment']}, ignore_index=True)

In [16]:
dataframe.head(10)

review sentiment
0  in 1974 the teenager martha moxley maggie grac...         1
1  ok so i really like kris kristofferson and his...         0
2  ***spoiler*** do not read this if you think ab...         0
3  hi for all the people who have seen this wonde...         1
4  i recently bought the dvd forgetting just how ...         0
5  leave it to braik to put on a good show finall...         1
6  nathan detroit frank sinatra is the manager of...         1
7  to understand crash course in the right contex...         1
8  ive been impressed with chavezs stance against...         1
9  this movie is directed by renny harlin the fin...         1

In [17]:
dataframe.to_csv('reviews_processed.csv')

# Make data frequency

In [18]:
dictionary = {}

In [19]:
import re
for i, row in dataframe.iterrows():
    texto = re.split(' ', row['review'])
    for word in texto:
        if len(word) > 0:
            try:
                if row['sentiment'] == 1:
                    count = dictionary[word][1]
                    dictionary[word][1] = count +1
                elif row['sentiment'] == 0:
                    count = dictionary[word][0]
                    dictionary[word][0] = count + 1
            except KeyError:
                if row['sentiment'] == 1:
                    dictionary[word] = [0, 1]
                elif row['sentiment'] == 0:
                    dictionary[word] = [1, 0]

## Get the dictionary of all words

In [20]:
dictionary

{'in': [87329, 98978],
 '1974': [56, 74],
 'the': [325483, 339782],
 'teenager': [223, 234],
 'martha': [80, 128],
 'moxley': [2, 7],
 'maggie': [56, 106],
 'grace': [322, 343],
 'moves': [410, 605],
 'to': [136688, 131195],
 'high': [2182, 2142],
 'class': [775, 940],
 'area': [327, 344],
 'of': [137230, 152038],
 'belle': [50, 49],
 'haven': [11, 15],
 'greenwich': [8, 24],
 'connecticut': [10, 62],
 'on': [34213, 33647],
 'mischief': [14, 18],
 'night': [2031, 2130],
 'eve': [116, 126],
 'halloween': [256, 219],
 'she': [11024, 12998],
 'was': [52183, 43272],
 'murdered': [299, 237],
 'backyard': [50, 22],
 'her': [15299, 19351],
 'house': [2047, 1974],
 'and': [146709, 175470],
 'murder': [839, 1219],
 'remained': [83, 93],
 'unsolved': [12, 17],
 'twenty': [332, 285],
 'two': [6360, 7131],
 'years': [3430, 5440],
 'later': [1726, 2522],
 'writer': [1193, 935],
 'mark': [540, 681],
 'fuhrman': [3, 4],
 'christopher': [309, 425],
 'meloni': [1, 17],
 'who': [18723, 21575],
 'is': [9

In [22]:
# Len of the dictionary, how many words was cached!

# Filter non valuable words

In [23]:
discriminatory_dictionary = dictionary.copy()

### Filter the dictionary

* First filter sum < 50

In [24]:
print('Initial Lenght>', len(discriminatory_dictionary))
for key in dictionary:
    L = dictionary[key]
    suma = L[0]+L[1]
    if suma < 50:
        del discriminatory_dictionary[key]
        
print('Final Lenght>', len(discriminatory_dictionary))

Initial Lenght> 139990
Final Lenght> 11273


* Second filter: percentege between 0.40 and 0.60

In [25]:
# Update the iterative dictionary
dictionary = discriminatory_dictionary.copy()

In [26]:
print('Initial Lenght>', len(discriminatory_dictionary))
for key in dictionary:
    L = dictionary[key]
    suma = L[0]+L[1]
    percent = L[0]/suma
    if percent < 0.6 and percent > 0.40:
        del discriminatory_dictionary[key]
        
print('Final Lenght>', len(discriminatory_dictionary))

Initial Lenght> 11273
Final Lenght> 5442


* Third Filter, 0 values

In [129]:
dictionary = discriminatory_dictionary.copy()

In [40]:
#print('Initial Lenght>', len(discriminatory_dictionary))
#for key in dictionary:
#    L = dictionary[key]
#    if L[0] == 0 or L[1] == 0:
#        del discriminatory_dictionary[key]
#        
#print('Final Lenght>', len(discriminatory_dictionary))

In [27]:
discriminatory_dictionary

{'martha': [80, 128],
 'maggie': [56, 106],
 'connecticut': [10, 62],
 'backyard': [50, 22],
 'years': [3430, 5440],
 'detective': [317, 495],
 'disgrace': [104, 23],
 'simpson': [93, 38],
 'trial': [110, 178],
 'moved': [218, 431],
 'andrew': [89, 140],
 'purpose': [546, 327],
 'writing': [1529, 980],
 'welcome': [150, 229],
 'support': [298, 457],
 'retired': [55, 105],
 'carroll': [21, 40],
 'robert': [789, 1193],
 'money': [3098, 1360],
 'cover': [786, 399],
 'true': [1657, 2835],
 'fifteen': [159, 94],
 'mother': [1185, 1795],
 'kennedy': [101, 173],
 'powerful': [275, 955],
 'family': [2119, 3571],
 'influence': [146, 264],
 'convicted': [47, 72],
 'hideous': [144, 44],
 'crime': [591, 895],
 'shows': [1871, 3025],
 'parallel': [76, 126],
 'lack': [1349, 757],
 'brazil': [71, 178],
 'available': [274, 582],
 'ok': [1499, 586],
 'easy': [586, 995],
 'low': [2351, 1211],
 'effortlessly': [25, 57],
 'holy': [131, 87],
 'bad': [14595, 3709],
 'kudos': [71, 181],
 'trying': [2974, 192

# Creating the dataframe

In [28]:
dataf = pd.DataFrame(columns=['word', 'probneg','probpos','freq'])

In [29]:
for key in discriminatory_dictionary:
    L = discriminatory_dictionary[key]
    suma = L[0]+L[1]
    dataf = dataf.append({'word':key, 'probneg':L[0]/suma, 'probpos':L[1]/suma,'freq':L[0]+L[1]}, ignore_index=True)

dataf.index.name = 'Id'

In [30]:
dataf.to_csv('frequency_probs.csv')

In [31]:
dataf.head(20)

word   probneg   probpos  freq
Id                                       
0        martha  0.384615  0.615385   208
1        maggie  0.345679  0.654321   162
2   connecticut  0.138889  0.861111    72
3      backyard  0.694444  0.305556    72
4         years  0.386697  0.613303  8870
5     detective  0.390394  0.609606   812
6      disgrace  0.818898  0.181102   127
7       simpson  0.709924  0.290076   131
8         trial  0.381944  0.618056   288
9         moved  0.335901  0.664099   649
10       andrew  0.388646  0.611354   229
11      purpose  0.625430  0.374570   873
12      writing  0.609406  0.390594  2509
13      welcome  0.395778  0.604222   379
14      support  0.394702  0.605298   755
15      retired  0.343750  0.656250   160
16      carroll  0.344262  0.655738    61
17       robert  0.398083  0.601917  1982
18        money  0.694930  0.305070  4458
19        cover  0.663291  0.336709  1185

# Get the positive features

In [36]:
PosFeatures = dataf.sort_values(['probpos', 'freq', 'probneg'], ascending=[0,0,1])
PosFeatures.reset_index(inplace = True, drop = True)
PosFeatures.index.name = 'Id'
PosFeatures.head(15)


word   probneg   probpos freq
Id                                    
0       ponyo  0.000000  1.000000  133
1        prot  0.000000  1.000000   81
2       gunga  0.000000  1.000000   60
3        prue  0.000000  1.000000   60
4       yokai  0.000000  1.000000   59
5        gypo  0.000000  1.000000   53
6      sosuke  0.000000  1.000000   51
7   blandings  0.000000  1.000000   51
8         din  0.012500  0.987500   80
9     antwone  0.013333  0.986667   75
10     gundam  0.020833  0.979167   96
11        dev  0.031746  0.968254   63
12    riddick  0.040000  0.960000   50
13     iturbi  0.040000  0.960000   50
14       edie  0.043103  0.956897  116

In [49]:
posProbArr = PosFeatures['probpos'].values
posFreqArr = PosFeatures['freq'].values
L = list(map(lambda x, y: x*y, posProbArr,posFreqArr))

len(L)
tempDF = pd.DataFrame(data={'PosValue':L})

PosFeaturesExtend = PosFeatures.join(tempDF)
PosFeaturesExtend = PosFeaturesExtend.sort_values(['PosValue'], ascending=False)
PosFeaturesExtend.reset_index(inplace = True, drop = True)
PosFeaturesExtend.index.name = 'Id'
PosFeaturesExtend.head(15)

word   probneg   probpos   freq  PosValue
Id                                             
0    great  0.286174  0.713826  18017   12861.0
1     well  0.399971  0.600029  20984   12591.0
2     also  0.399273  0.600727  17870   10735.0
3     even  0.613747  0.386253  24704    9542.0
4       no  0.634085  0.365915  25014    9153.0
5     love  0.331549  0.668451  12891    8617.0
6     best  0.325209  0.674791  12555    8472.0
7     life  0.371390  0.628610  12674    7967.0
8     dont  0.605058  0.394942  16924    6684.0
9      any  0.606408  0.393592  15013    5909.0
10   years  0.386697  0.613303   8870    5440.0
11  acting  0.627756  0.372244  12747    4745.0
12    plot  0.632437  0.367563  12874    4732.0
13   world  0.362662  0.637338   7183    4578.0
14   young  0.364248  0.635752   6987    4442.0

In [50]:
PosFeatures.to_csv('Extract_positive_Features.csv')
PosFeaturesExtend.to_csv('Extract_positive_Features2.csv')

# Get the negative features

In [52]:
NegFeatures = dataf.sort_values(['probneg'], ascending=False)
NegFeatures.reset_index(inplace = True, drop = True)
NegFeatures.index.name = 'Id'
NegFeatures.head(15)

word   probneg   probpos freq
Id                                     
0       komodo  1.000000  0.000000   61
1    carnosaur  1.000000  0.000000   61
2      piranha  1.000000  0.000000   50
3   hobgoblins  1.000000  0.000000   53
4        manos  1.000000  0.000000   61
5           *1  0.993865  0.006135  163
6          uwe  0.993464  0.006536  153
7         boll  0.989305  0.010695  187
8   chupacabra  0.986667  0.013333   75
9     caligula  0.972973  0.027027   74
10     killjoy  0.968750  0.031250   64
11  forwarding  0.965517  0.034483   58
12   henriksen  0.961538  0.038462   52
13        ajay  0.960000  0.040000   50
14     beowulf  0.958333  0.041667   72

In [53]:
NegFeatures.to_csv('Extract_negative_Features.csv')

# Result
Here is Showed the most important words to be analysed it, depends on its apearances and it's percentege
Over 20 Words of each class is selected

* Extreme Positive Words:

TO BOOLEAN FEATURE:

|Word | Probability Pos | Frequency|
|----- | ----- | -----|
|excellent| 	0.818070  |	4073|
|terrific |	0.81664   |	829|
|perfection|	0.8850	|	287|
|wonderfully|	0.8836	|	593|
|captures |	0.875	|	400|
|matthau	|	0.94560	|	239|
|custer	|	0.95625	|	160|
|felix	|	0.9456	|	239|
|paulie	|	0.936	|	110|
|miyazaki|	0.93388	|	121|
|devito	|	0.925	|	108|
|haines	|	0.914599|	117|
|gromit	|	0.91	|	100|
|flawless|	0.894	|	236|
|superbly|	0.889	|	244|
|understated|	0.8864	|	185|
|perfection|	0.885	|	287|
|lemmon	|	0.881	|	161|
|heartbreaking|	0.87	|	154|
|delightful|	0.867	|	469	|
|beautifully|	0.864	|	868|
|timeless|	0.863	|	234|
|underrated|	0.8614	|	462|
|cassavetes|	0.859	|	192|
|superb	|	0.857	|	1293|
|fantastic|	0.808	|	1494	|
|delight	|	0.8098	|	305|
|bruno	|	0.8104	|	211|
|kitty	|	0.81168	|	154|
|brilliantly|	0.854	|	466|
|extraordinary|	0.8469	|	392|
|outstanding|	0.8394	|	785|
|powell	|	0.837	|	259|
|pleasantly|	0.837	|	234|
|touching|	0.8333	|	810|
|finest	|	0.831	|	529|
|friendship|	0.831	|	581|
|poignant|	0.829	|	304|
|marvelous|	0.828	|	291|
|wonderful|	0.827	|	3200|
|magnificent|	0.821	|	539|
|terrific|	0.8166	|	829|


* Extreme Negative Words:

TO BE BOOLEAN:

|Word       |Percentege |Freq|
| -------   | --------- | ---|
|bad		|0.797		|1830|
|dumb		|0.80		|1222|
|annoying	|0.806		|1912|
|ridiculous	|0.81		|1873|
|fails		|0.812		|1160|
|boring		|0.822		|3586|
|avoid		|0.824		|1496|
|dull		|0.83		|1587|
|badly		|0.834		|1272|
|stupid		|0.84		|3474|
|mess		|0.842		|1233|
|worse		|0.844		|2908|
|crap		|0.855		|2032|
|horrible	|0.864		|2496|
|terrible	|0.870		|3292|
|wasted		|0.871		|1163|
|lame		|0.883		|1378|
|poorly		|0.903		|1387|
|awful		|0.912		|3409|
|worst		|0.916		|5313|
|waste		|0.936		|2783|

64280 appereances

* With less than 1000 & greather than 700 and greather than 0.8 of probabilities:

|Word       |Percentege |Freq|
| -------   | --------- | ---|
|bother		|0.811		|890
|excuse		|0.817		|799
|garbage	|0.859	    |952
|pathetic	|0.872		|973
|pointless	|0.893		|974
|laughable	|0.905		|819

* With less than 700 & greather than 550 and greather than 0.8 of probabilities:

|Word       |Percentege |Freq|
| -------   | --------- | ---|
|wooden		|0.832		|649|
|whatsoever	|0.847		|650|
|redeeming	|0.91		|614|



